### PR TITLE
Rework ImageFileDialog URLs

### LINKED ISSUE #1389: [ImageFileDialog] Rework absolute/relative URI

### LINKED ISSUE #1390: [ImageFileDialog] ReworkURI using silx.io.uri
- Provide an URI class
- Provide an API to reach the data from the URI

### PR DESCRIPTION
**Pull request by vallsv**
Here is an implementation of a `DataUrl`.

Closes #1389
Closes #1390 

- It was designed to reach data from `fabio` or `silx.io.open` with support of slicing.
- It can parse regular files
- It supports absolute (`silx:///data.h5`) and relative (`silx:data.h5`) file path
- It supports Windows drives

Here is few examples, but you better to check the tests:
```
# fabio access using absolute path
DataUrl("fabio:///data/image.edf::[2]")
DataUrl("fabio:///C:/data/image.edf::[2]")

# silx access using absolute path
DataUrl("silx:///data/image.h5::/data/dataset[1,5]")
DataUrl("silx:///data/image.edf::/scan_0/detector/data")
DataUrl("silx:///C:/data/image.edf::/scan_0/detector/data")

# Relative path access
DataUrl("silx:./image.h5")
DataUrl("fabio:./image.edf")
DataUrl("silx:image.h5")
DataUrl("fabio:image.edf")

# Is also support parsing of file access for convenience
DataUrl("./foo/bar/image.edf")
DataUrl("C:/data/")
```

It is a redesign of the `_ImageUrl` class available on `silx.gui.dialog.ImageFileDialog`.
- To have something closer to the default URLs (especially for absolute file paths)
- Something which can be used with other widgets like #1395 
- And outside the widgets with a dedicated `silx.io` API.

### CODE REVIEW

**Comment by t20100** [comment]
This sounds good to me.

I would be good to add the doc for `DataUrl.__init__` arguments (even if it is obvious).
For consistency, I would consider having the same name for the `slicing` `__init__` argument and the `slice` method (`slicing` or `data_slice`?).

**Comment by vallsv** [comment]
Then i rename the slice as suggested `data_slice`, and i update the documentation.

Now the PR also contains `silx.io.get_data`.

I start working on reusing `DataUrl` with the `ImageFileDialog` in order to be able to reuse URL from the dialog as an input of `silx.io.get_data`.

**Comment by vallsv** [comment]
This part is fine. Ready for a review.

It then creates an API `silx.io.get_data`, and remove `_ImageUri`.

We can talk about the URL next time. This should not change too much things for now, cause everything is using `DataUrl` now.

File: silx/io/utils.py [hunk 1]
```diff
@@ -753,9 +757,89 @@ def is_softlink(obj):
...
+            raise ValueError("Fabio slice expect a single frame, but %s found" % data_slice)
+        index = data_slice[0]
+        if not isinstance(index, int):
+            raise ValueError("Fabio slice expect a single integer, but %s found" % data_slice)
+
+        try:
+            fabio_file = fabio.open(url.file_path())
+        except Exception:
+            logger.debug("Error while opening %s with fabio", url.file_path(), exc_info=True)
+            IOError("Error while opening %s with fabio (use debug for more information)" % url.path())
```
Comment (t20100) [inline silx/io/utils.py]: Looks like a `raise` is missing

Comment (vallsv) [reply silx/io/utils.py]: Thanks

**Pull request merged**